### DISCUSSION TITLE
Big Refactor

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/persisted.nvim#145] [Announcements]
## tl;dr

- I've broken my own [mantra](https://github.com/olimorris/persisted.nvim/blob/main/CONTRIBUTING.md) in recent times
- So I've [refactored](https://github.com/olimorris/persisted.nvim/pull/144) the plugin to restore it back to my goal of "simple session management"
- The impact is that some smaller features have been removed
- But...the plugin is now completely extensible
- And...the startup time is noticeably faster

## What’s the impact?

Config items and/or features that have been removed:

- `default_branch` removed
- `branch_separator` removed (I've searched GitHub and reached out to the two users it may impact)
- `silent` removed
- `telescope.reset_prompt` removed
- `autosave` becomes `autostart` (backwards compatible)
- `should_autosave` becomes `should_save` (backwards compatible)
- `PersistedStateChange` hook has been renamed to `PersistedStart` and `PersistedStop` (thanks @neandrake)
- `ignored_branches` removed
- Autoloading when args are passed to Neovim, removed

Regarding the removal of autoloading with args (7944b9a) and ignored branches (ce9d621); this added a lot of bloat to the plugin and is something that is relatively trivial to add to a user's config. As a result, I've [outlined](https://github.com/olimorris/persisted.nvim#building_construction-extending-the-plugin) how to accomplish both of these in the README.

I've made all the functions in the `init.lua` and `utils.lua` files public. This allows you to extend the plugin to suit your needs without having to merge any PRs (hopefully!). Setting `autostart = false`, a user can perform some logic and other steps before using `persisted.start()`.

## Future deprecations

A heads up for people who have copied the plugin's config and placed it in their config...soon, the following settings will cause an error:

```lua
allowed_dirs = nil,
ignored_dirs = nil,
```

To resolve this, simply delete them or set them to `{}`.

**Comment by adrian5** [comment]
Seems like a sound decision Oli! I take it `allowed/ignored_dirs` is to be handled in `should_save`?

**Reply by olimorris** [reply]
> Seems like a sound decision Oli! I take it `allowed/ignored_dirs` is to be handled in `should_save`?

They will automatically be handled for you via the `autoload` method if you have the setting enabled. Alternatively, there is a method called `allowed_dir` that returns a boolean if the current working directory is in the allowed list and not in the ignored one.

Let me know if you're having any trouble getting it to work.

**Reply by adrian5** [reply]
Sorry to bump this old thread Oli. I thought I had this working but was mistaken. I'd like to only save sessions for `cwd`s that are sub-directories of `allowed_dirs`. I now started with a minimal function using `allowed_dir` but that seems to save any session:

```lua
should_save = function()
  local allowed_dir = require("persisted").allowed_dir -- Am I importing it right?
  return allowed_dir()
end,
allowed_dirs = {
  "~/code",
  "~/notes/api"
}
```
Thanks!

**Reply by olimorris** [reply]
I'd probably do something like:

```lua
{
  "olimorris/persisted.nvim",
  lazy = false,
  config = function()
    local persisted = require("persisted")
    local utils = require("persisted.utils")
    local allowed_dirs = {
      "~/code",
      "~/notes/api"
    }

    persisted.setup({
      should_save = function()
        return utils.dirs_match(vim.fn.getcwd(), allowed_dirs)
      end,
    })
  end,
}
```

I haven't tested this but you hopefully get the gist. Also the `dirs_match` may not be adequate for your use case.

**Reply by adrian5** [reply]
That seems to work just fine Oli, I tried with different dirs. Thanks!

**Comment by neandrake** [comment]
Following the guide you linked, this is what I have to achieve the previous behavior,

```lua
-- Persisted does not autoload unless nvim is opened with a file. This
-- will force auto-loading if nvim has no arguments or is passed a
-- single argument that is a directory, otherwise persisted will kick in
-- like normal.
vim.api.nvim_create_autocmd("VimEnter", {
    nested = true,

    callback = function()
        if vim.g.started_with_stdin then
            return
        end

        local forceload = false
        if vim.fn.argc() == 0 then
            forceload = true
        elseif vim.fn.argc() == 1 then
            local dir = vim.fn.expand(vim.fn.argv(0))
            if dir == '.' then
                dir = vim.fn.getcwd()
            end

            if vim.fn.isdirectory(dir) ~= 0 then
                forceload = true
            end
        end

        persisted.autoload({ force = forceload })
    end,
})
```

**Reply by olimorris** [reply]
Awesome thanks. I'll add this to the README and shout you out.

**Comment by neandrake** [comment]
@olimorris could you also include in your description above that the `PersistedStateChange` hook has been split/renamed to `PersistedStart` and `PersistedStop`?

**Reply by olimorris** [reply]
Great shout and thank you
